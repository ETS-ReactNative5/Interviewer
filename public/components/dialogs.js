const { dialog } = require('electron');

const openDialogOptions = {
  buttonLabel: 'Open',
  nameFieldLabel: 'Open:',
  defaultPath: 'Protocol.netcanvas',
  filters: [{ name: 'Network Canvas', extensions: ['netcanvas'] }],
  properties: ['openFile'],
};

const openDialog = () =>
  new Promise((resolve, reject) => {
    dialog.showOpenDialog(openDialogOptions, (filenames) => {
      if (!filenames) { reject('Import protocol dialog cancelled.'); return; }
      if (!filenames.length || filenames.length !== 1) { reject('Only a single protocol may be imported at a time.'); return; }
      resolve(filenames[0]);
    });
  });

module.exports = {
  openDialog,
};
