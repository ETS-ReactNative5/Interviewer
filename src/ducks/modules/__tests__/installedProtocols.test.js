/* eslint-env jest */
/* eslint-disable @codaco/spellcheck/spell-checker */

import { createStore, applyMiddleware } from 'redux';
import thunks from 'redux-thunk';

import reducer, { actionCreators, actionTypes } from '../protocols';
import deleteProtocol from '../../../utils/protocol/deleteProtocol';
import { actionCreators as dialogsActions } from '../dialogs';

jest.mock('../dialogs');
jest.mock('../../../utils/protocol/deleteProtocol');

const testMiddleware = (actionListener) => () => (next) => (action) => {
  actionListener(action);
  return next(action);
};

const getStore = (actionListener, initialState) => (
  createStore(reducer, initialState, applyMiddleware(thunks, testMiddleware(actionListener)))
);

describe('protocols reducer', () => {
  it('should return the initial state', () => {
    expect(
      reducer(undefined, {}),
    ).toEqual({});
  });

  describe('DELETE_PROTOCOL', () => {
    const protocolUID = '1234';

    const action = {
      type: actionTypes.DELETE_PROTOCOL,
      protocolUID,
    };

    const result = reducer(
      {
        [protocolUID]: { uid: protocolUID },
      },
      action,
    );

    expect(result).toEqual({});
  });

  describe('deleteProtocol()', () => {
    const protocolUID = '1234';

    beforeEach(() => {
      dialogsActions.openDialog.mockReset();
      deleteProtocol.mockReset();
    });

    it('confirms and then deletes protocol', (done) => {
      const actionListener = jest.fn();

      dialogsActions.openDialog.mockImplementationOnce(
        () => () => Promise.resolve(true),
      );

      const store = getStore(actionListener);
      store.dispatch(actionCreators.deleteProtocol(protocolUID));

      setImmediate(() => {
        expect(actionListener.mock.calls).toMatchSnapshot();
        expect(deleteProtocol.mock.calls.length).toBe(1);
        done();
      });
    });

    it('does not confirm and does not delete protocol', (done) => {
      const actionListener = jest.fn();

      dialogsActions.openDialog.mockImplementationOnce(
        () => () => Promise.resolve(false),
      );

      const store = getStore(actionListener);
      store.dispatch(actionCreators.deleteProtocol(protocolUID));

      setImmediate(() => {
        expect(actionListener.mock.calls).toMatchSnapshot();
        expect(deleteProtocol.mock.calls.length).toBe(0);
        done();
      });
    });

    describe('Has existing sessions', () => {
      it('warns user about existing sessions', (done) => {
        const actionListener = jest.fn();

        dialogsActions.openDialog.mockImplementationOnce(
          () => () => Promise.resolve(false),
        );

        const store = getStore(actionListener, {
          sessions: [
            { protocolUID, lastExportedAt: Date.now() },
          ],
        });
        store.dispatch(actionCreators.deleteProtocol(protocolUID));

        setImmediate(() => {
          expect(dialogsActions.openDialog.mock.calls).toMatchSnapshot();
          expect(deleteProtocol.mock.calls.length).toBe(0);
          done();
        });
      });

      it('warns user about existing sessions, but allows deletion on confirm', (done) => {
        const actionListener = jest.fn();

        dialogsActions.openDialog.mockImplementationOnce(
          () => () => Promise.resolve(true),
        );

        const store = getStore(actionListener, {
          sessions: [
            { protocolUID, lastExportedAt: Date.now() },
          ],
        });
        store.dispatch(actionCreators.deleteProtocol(protocolUID));

        setImmediate(() => {
          expect(deleteProtocol.mock.calls.length).toBe(1);
          done();
        });
      });
    });

    describe('Has non-exported existing sessions', () => {
      it('warns user about existing sessions', (done) => {
        const actionListener = jest.fn();

        dialogsActions.openDialog.mockImplementationOnce(
          () => () => Promise.resolve(false),
        );

        const store = getStore(actionListener, {
          sessions: [
            { protocolUID },
          ],
        });
        store.dispatch(actionCreators.deleteProtocol(protocolUID));

        setImmediate(() => {
          expect(dialogsActions.openDialog.mock.calls).toMatchSnapshot();
          expect(deleteProtocol.mock.calls.length).toBe(0);
          done();
        });
      });

      it('warns user about existing sessions, but allows deletion on confirm', (done) => {
        const actionListener = jest.fn();

        dialogsActions.openDialog.mockImplementationOnce(
          () => () => Promise.resolve(true),
        );

        const store = getStore(actionListener, {
          sessions: [
            { protocolUID },
          ],
        });
        store.dispatch(actionCreators.deleteProtocol(protocolUID));

        setImmediate(() => {
          expect(deleteProtocol.mock.calls.length).toBe(1);
          done();
        });
      });
    });
  });
});
