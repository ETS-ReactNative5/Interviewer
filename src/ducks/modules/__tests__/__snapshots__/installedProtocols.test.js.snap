// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protocols reducer deleteProtocol() Has existing sessions warns user about existing sessions 1`] = `
Array [
  Array [
    Object {
      "confirmLabel": "Delete protocol and sessions",
      "message": <React.Fragment>
        There are sessions that use this protocol.
        <br />
        <br />
        <strong>
          Deleting this protocol will also delete associated sessions.
        </strong>
      </React.Fragment>,
      "title": "Session(s) using protocol",
      "type": "Confirm",
    },
  ],
]
`;

exports[`protocols reducer deleteProtocol() Has non-exported existing sessions warns user about existing sessions 1`] = `
Array [
  Array [
    Object {
      "confirmLabel": "Delete protocol and sessions",
      "message": <React.Fragment>
        There are sessions that use this protocol which have not yet been exported.
        <br />
        <br />
        <strong>
          Deleting this protocol will also delete associated sessions.
        </strong>
      </React.Fragment>,
      "title": "Non-exported session(s) using protocol",
      "type": "Warning",
    },
  ],
]
`;

exports[`protocols reducer deleteProtocol() confirms and then deletes protocol 1`] = `
Array [
  Array [
    Object {
      "protocolUID": "1234",
      "type": "INSTALLED_PROTOCOLS/DELETE_PROTOCOL",
    },
  ],
]
`;

exports[`protocols reducer deleteProtocol() does not confirm and does not delete protocol 1`] = `Array []`;
